clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -420,7 +420,7 @@ private static int parseAcks(String acksString) {
      *
      * @throws InterruptException If the thread is interrupted while blocked
      * @throws SerializationException If the key or value are not valid objects given the configured serializers
-     * @throws BufferExhaustedException If <code>block.on.buffer.full=false</code> and the buffer is full.
+     * @throws TimeoutException if the time taken for fetching metadata, serialization of (key,value), partitioning has surpassed <code>max.block.ms</code>.
```

Comment:
`doSend` still has a catch block for `BufferExhaustedException`, but it doesn't look like we throw that any longer. We should probably fix that too.

Since `BufferExhaustedException` is not used anywhere, I wonder if we should deprecate it? That should probably be a separate JIRA though.

Finally, are you considering the following change that added "Blocking in the user-supplied serializers or partitioner will not be counted against this timeout" to `MAX_BLOCK_MS_DOC` for performance reasons?

https://github.com/apache/kafka/pull/632/commits/69d94701d4de3283f3783826edb9321dc6c800ec